File: src/main/java/com/bvanseg/gigeresque/common/fluid/Fluids.java
package com.bvanseg.gigeresque.common.fluid;

import com.bvanseg.gigeresque.common.Gigeresque;
import com.bvanseg.gigeresque.common.util.GigeresqueInitializer;
import com.bvanseg.gigeresque.common.util.InitializationTimer;

import net.minecraft.fluid.Fluid;
import net.minecraft.util.Identifier;
import net.minecraft.util.registry.Registry;

public class Fluids implements GigeresqueInitializer {
	private Fluids() {
	}

	private static Fluids instance;

	synchronized public static Fluids getInstance() {
		if (instance == null) {
			instance = new Fluids();
		}
		return instance;
	}

	private <T extends Fluid> T registerFluid(String path, T fluid) {
		return Registry.register(Registry.FLUID, new Identifier(Gigeresque.MOD_ID, path), fluid);
	}

	public static final BlackFluid BLACK_FLUID_STILL = new BlackFluid.Still();
	public static final BlackFluid BLACK_FLUID_FLOWING = new BlackFluid.Flowing();

	@Override
	public void initialize() {
		InitializationTimer.initializingBlock("Fluids", this::initializeImpl);
	}

	private void initializeImpl() {
		registerFluid("black_fluid_still", BLACK_FLUID_STILL);
		registerFluid("black_fluid_flowing", BLACK_FLUID_FLOWING);
	}
}

File: src/main/kotlin/com/bvanseg/gigeresque/common/fluid/Fluids.kt
//package com.bvanseg.gigeresque.common.fluid
//
//import com.bvanseg.gigeresque.common.Gigeresque
//import com.bvanseg.gigeresque.common.util.GigeresqueInitializer
//import com.bvanseg.gigeresque.common.util.initializingBlock
//import net.minecraft.fluid.Fluid
//import net.minecraft.util.Identifier
//import net.minecraft.util.registry.Registry
//
///**
// * @author Boston Vanseghi
// */
//object Fluids : GigeresqueInitializer {
//
//    private fun <T : Fluid> registerFluid(path: String, fluid: T): T =
//        Registry.register(Registry.FLUID, Identifier(Gigeresque.MOD_ID, path), fluid)
//
//    val BLACK_FLUID_STILL = BlackFluid.Still()
//    val BLACK_FLUID_FLOWING = BlackFluid.Flowing()
//
//    override fun initialize() = initializingBlock("Fluids") {
//        registerFluid("black_fluid_still", BLACK_FLUID_STILL)
//        registerFluid("black_fluid_flowing", BLACK_FLUID_FLOWING)
//    }
//}